In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Modeling & Evaluation

## Objective
Train and evaluate machine learning models to predict passenger survival
on the Titanic dataset. Multiple models are compared to select the best
performing approach.


### import libraries

In [4]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score ,confusion_matrix, classification_report

In [5]:
data=pd.read_csv("/kaggle/input/processed-data/train_cleaned.csv")
data.head(10)


Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_Q  Embarked_S  \
0         0       3    0  22.0      1      0   7.2500       False        True   
1         1       1    1  38.0      1      0  71.2833       False       False   
2         1       3    1  26.0      0      0   7.9250       False        True   
3         1       1    1  35.0      1      0  53.1000       False        True   
4         0       3    0  35.0      0      0   8.0500       False        True   
5         0       3    0  28.0      0      0   8.4583        True       False   
6         0       1    0  54.0      0      0  51.8625       False        True   
7         0       3    0   2.0      3      1  21.0750       False        True   
8         1       3    1  27.0      0      2  11.1333       False        True   
9         1       2    1  14.0      1      0  30.0708       False       False   

   FamilySize  
0           2  
1           2  
2           1  
3           2  
4           1  
5           1  
6           1  
7           5  
8           3  
9           2

## Define Features (X) and Target (y)

In [6]:
X=data.drop(columns=["Survived"])
Y=data["Survived"]

## Train-validation split

The dataset is split into training and validation sets to evaluate model
performance on unseen data and avoid overfitting.


In [7]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,random_state=42)

## Model 1: Logistic Regression

Logistic Regression is used as a baseline model due to its simplicity
and interpretability.


In [8]:
model=LogisticRegression(max_iter=1000)
model.fit(X_train,Y_train)
Y_pred=model.predict(X_test)

## Evaluate Logistic Regression

In [9]:
print("Logistic Regression Accuracy:",accuracy_score(Y_test,Y_pred))
print(confusion_matrix(Y_test,Y_pred))
print(classification_report(Y_test,Y_pred))

Logistic Regression Accuracy: 0.8100558659217877
[[90 15]
 [19 55]]
              precision    recall  f1-score   support

           0       0.83      0.86      0.84       105
           1       0.79      0.74      0.76        74

    accuracy                           0.81       179
   macro avg       0.81      0.80      0.80       179
weighted avg       0.81      0.81      0.81       179



## Model 2: Random Forest

Random Forest is an ensemble model capable of capturing non-linear
relationships and feature interactions.


In [12]:
rf=RandomForestClassifier(n_estimators=100,random_state=42)
rf.fit(X_train,Y_train)
Y_pred_rf=rf.predict(X_test)

In [14]:
print("Random Forest accuracy:",accuracy_score(Y_test,Y_pred_rf))
print(confusion_matrix(Y_test,Y_pred_rf))
print(classification_report(Y_test,Y_pred_rf))

Random Forest accuracy: 0.8212290502793296
[[91 14]
 [18 56]]
              precision    recall  f1-score   support

           0       0.83      0.87      0.85       105
           1       0.80      0.76      0.78        74

    accuracy                           0.82       179
   macro avg       0.82      0.81      0.81       179
weighted avg       0.82      0.82      0.82       179



## Evaluate Random Forest

The Random Forest model achieved an accuracy of 82% on the validation set.
The confusion matrix shows balanced performance across both classes, with
slightly better performance in predicting non-survivors. Precision and recall
values indicate that the model generalizes well without significant overfitting.
